lib/luacrypto.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * SPDX-FileCopyrightText: (c) 2025 jperon <cataclop@hotmail.com>
+ * SPDX-License-Identifier: MIT OR GPL-2.0-only
+ */
+
+#ifndef _LUACRYPTO_H
+#define _LUACRYPTO_H
+
+#include <linux/err.h>
+#include <lua.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+/**
+ * LUACRYPTO_NEW - Macro to define a common constructor for crypto transform objects.
+ * @name: The base name of the crypto module (e.g., aead, skcipher, shash).
+ * @T: The C struct type of the crypto transform (e.g., struct crypto_aead).
+ * @alloc: The kernel crypto allocation function (e.g., crypto_alloc_aead).
+ * @class: The lunatik_class_t instance for this object type.
+ * @private: The expression to assign to object->private (e.g., tfm or sdesc).
+ * @extra: Optional extra allocation and assignment logic (e.g., for shash_desc).
+ *
+ * This macro generates the `luacrypto_<_name>_new` function, which acts as the
+ * constructor for Lua crypto objects. It handles the common steps of checking
+ * the algorithm name, allocating the Lunatik object, allocating the kernel
+ * crypto transform, handling errors, and assigning the private data.
+ */
+#define LUACRYPTO_NEW(name, T, alloc, class, priv, ...)						\
+static int luacrypto_##name##_new(lua_State *L)								\
+{													\
+	const char *algname = luaL_checkstring(L, 1);							\
+	lunatik_object_t *object = lunatik_newobject(L, &class, 0);					\
+													\
+	T *tfm = alloc(algname, 0, 0);									\
+	if (IS_ERR(tfm)) {										\
+		long err = PTR_ERR(tfm);								\
+		luaL_error(L, "Failed to allocate " #name " transform for %s (err %ld)", algname, err);	\
+	}												\
+	__VA_ARGS__											\
```

Comment:
It looks a weird usage of varagargs.. I've made my point in another comment.. as general rule, I don't think it is a good idea to use any macro's argument like this.. I think it's always better to threat like functions if it's not an assignment..

Reply:
If so, IMHO it would be better to keep this macro without the last argument for common cases, and to write the full function in specific cases.

Reply:
But doing that would imply:
- either seeing again the `, )` at the end of `LUACRYPTO_NEW` calls when there is no extra part;
- or defining a "nop" function in those cases.
Is this what you intend?

Reply:
I ended with a solution using two macros for each, with a _CUSTOM version allowing to call a function for extra logic.